### ISSUE TITLE
Prevent Duplicate Jobs From Running In Load-Balanced Environment

### ISSUE THREAD

**Issue by pkmccaffrey** [fluentscheduler/FluentScheduler#167]
I'm using FluentScheduler as the basis of a job-scheduling system in an application.  My implementation uses a relational database as persistent storage for job configurations (timing, class types, etc.).  I'd like to further extend this database table to handle "job locking" in a load-balanced environment.

My plan was to attach to the JobStart event.  This handler would check if the database record that corresponds to a given IJob is currently being executed elsewhere.  If it is, the handler would prevent the job  from running.  However, I can't seem to find a way to prevent a job from running (or a way to kill a running job).  Does this functionality exist?

**Comment by tallesl** [comment]
Why not just checking such condition in the job itself (and just returning without doing nothing when necessary)?

If you don't want to mix both responsibilities in a single job (checking if it's suppose to do the work and actually doing the work), you can make two jobs: one that just checks and, if needed, starts the other.

You can start a job from outside the registry:

```cs
JobManager.AddJob<MyJob>(s => s.ToRunNow());
```

**Comment by pkmccaffrey** [comment]
@tallesl - I could certainly do as you suggested, but having support for load-balanced environments seems like it could be useful for others, and may be worth adding to the project itself.

I noticed [this feature request](https://github.com/fluentscheduler/FluentScheduler/issues/68), which requires reading/writing database records for each job (I'm already doing this).  Adding support for load-balanced environments would probably fit nicely with that feature.  I'm not sure if my implementation is generic enough, but I'd be happy to help if you're looking for any.

**Comment by tallesl** [comment]
About that feature request, it would be a useful one to have, but tricky to tackle.

Maybe you want to store it in a SQL database? Maybe NoSQL? XML files? JSON? YAML? What about the new fad technology around the corner?

I think the right way to address such issue is by offering an easy way of serialize/deserialize the library state and do no actual storage work. Even that approach is tricky to implement, right now the types involved on the scheduling are far from being serializable.

**Comment by pkmccaffrey** [comment]
> Maybe you want to store it in a SQL database? Maybe NoSQL? XML files? JSON? YAML? What about the new fad technology around the corner?

Personally, I think handling it in a similar fashion to IJobFactory would work out well.  For example, there could be an IJobStore which has methods for getting/saving jobs from whatever storage you want to use.  JobManager would have a JobStore property that the use sets to whatever IJobStore they want to use.

> right now the types involved on the scheduling are far from being serializable.

This is the biggest problem that I ran into.  The schema I came up with covers all of the use cases that I need, but not all of the ones that this library offers.
